### ISSUE TITLE
Experimental results on MLP projectors

### ISSUE THREAD

**Issue by ruiyicheng** [rbalestr-lab/lejepa#23]
Hi! According to the author's [suggestion ](https://github.com/rbalestr-lab/lejepa/issues/17#issuecomment-3579000750 ) #17  I performed some experiments on the effects of MLP projector and found some interesting results. 
I adapted the minimum example to enable tracking the per-layer linear probe accuracy while training. The loss is still on the final layer. I recorded the loss (just like the provided eample) for each epoch while using ViT-S8/ResNet-50 (both of them have ~20 m parameters).
For ViT-S8, I observe an overfitting (low loss but worse accuracy) on some setting of multi-layer MLP projection.The accuracy is always ~15% while not using any mlp projector(I abort it after 300 epoch, thus not shown here). The performance of ViT-8 is the best while using one-lay MLP projector. Interestingly, the results from the last hidden layer seems the worst for overfitting MLP
However, the results of ResNet-50 is much more stable. I can get a very good stable result even when not using any MLP projectors. 
 
The title of the figures mean {embedding output dim}_{MLP hidden dim}_{out dim}. All the losses are measured on out_dim.

I also show the singularity value distribution for different layer activation on test set for the model at final epoch

I am very interested in conducting more experiments on this topic. If you would like additional experiments or further analysis, please let me know!

Looking forward to hearing your thoughts.

[experimental_results.zip](https://github.com/user-attachments/files/23981492/experimental_results.zip)

**Comment by ruiyicheng** [comment]
For easier inspection, I show the images here.
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/9d263360-5146-4d00-b91f-5c4127fcc14f" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/edb1851d-69cb-4e26-9af2-c5273e06162d" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/ee3d23bd-6902-40b4-aa1c-f786f48fd671" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/56912532-3907-47d3-a180-7889fbc7ccaa" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/275d9758-8d44-4d79-8f2e-a9b51ac8eaf9" />
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/2d17ee53-c39d-45d5-8b2c-f23a39067709" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/46c408c1-4cfc-4bad-a976-5c01e6c73ad0" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/bf24b644-923c-444e-bc99-57ec82cba746" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/269dcaf6-7281-4fb6-be96-33dcf6d8d6f5" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/c720e8ad-cb1f-43cf-8c99-7deda51d9cdb" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/0f22412f-b6bb-4d79-b71d-f7ecde9632c5" />
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/39c1f2af-944c-475c-888d-e807a0205517" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/0421a436-9cd1-4656-995b-e040d91f3314" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/9718034d-3488-466d-8500-48091879285c" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/852ae60a-d5f7-4a76-a2fd-4f8cf1b68885" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/03a227a7-1d68-4f90-8210-576a7df22299" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/f5e0495e-d370-4ce0-98ac-3fb9070a52d3" />
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/e293b364-430b-4012-b3c2-351a18304e3c" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/1b4edbd8-2e9f-4303-8e02-90bb8ef5610e" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/2147ff6b-0b6e-4843-998b-73e2bf58bd7a" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/8c76245a-e369-41f6-868a-1c948bd430cf" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/528d0a4b-89b2-4148-830c-c7e79129a7f3" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/635c2a36-854c-453e-bc1a-69aab12b0ed8" />
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/8b75c5a7-4fc4-456f-9c35-8081a1c7ffa8" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/5179a68d-575f-439c-9b71-bb2b011a0774" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/852d660c-ac29-4d26-9ca3-54cdf7da1ed2" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/9e0c1073-b170-4fe4-8feb-3c50a97b6443" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/62744473-098c-4bdf-a41b-1ea026e5387d" />
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/3292f7d9-6e6e-4dcd-a4dd-ade7c1151795" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/21d707db-f695-40bb-8814-4036f473340d" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/76d3e868-de54-4bd2-a69b-7839c3e7ced3" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/eb7358f1-7939-4ff3-95ed-c188fb8a9a3d" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/5e0d8990-7475-42a4-aa41-8abed3dde236" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/8067565a-1ab4-49c8-8650-699d37d7fec9" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/d3c954c5-ab1e-42c6-ae4e-c669f8ce1f5c" />
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/1fe86a2b-bbb3-4ab1-ae02-ae5b65e3829e" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/cc65c80d-af7e-4722-875c-f139027aac2e" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/b1aa869d-c323-47ee-89dd-15b20dfccd72" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/66614399-16f0-462d-b3b4-70deb4b10fd9" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/4fb72bb1-3182-4b68-a970-34f637b5c050" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/f25111b5-1455-4bcc-bd74-53e5c7af120d" />
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/1c9dca9a-82fc-43f8-9a8a-95a07b200526" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/7965d0b7-6d8c-4f42-8893-6736fb940ca5" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/4a6abe8b-6615-4d02-a28d-8791f7eb4b7f" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/9ba6cd0e-6769-49a0-9816-b67175541acb" />
<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/fbd413d8-e5b4-4ec8-9c35-b9cb9584917b" />
<img width="640" height="480" alt="Image" src="https://github.com/user-attachments/assets/158f13eb-900e-4471-86bb-176e231c3809" />
<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/f43ae573-0a48-4203-9fd6-f961ba8a5f7c" />

**Comment by ruiyicheng** [comment]
@RandallBalestriero Dear professor, I have obtained the results with no MLP projectors on the imagenet-100 data. The accuracy of the linear probe with no MLP is still lower than Table 4 correspondings of the paper for all the model I have tried so far. However, the models seems still have room for higher accuracy at epoch=400 as used by paper, which can be due to the difference of hyperparameters.  Please let me know if more experiments are needed.

Meanwhile, I found in other experiments that using CLIP loss combined with LeJEPA loss would make SIGReg loss decrease faster in my task (alignment of multi-model view of the same light curve (brightness of star change with respect to time)). 

<img width="1000" height="600" alt="Image" src="https://github.com/user-attachments/assets/1318b80d-d5d4-4777-8b43-e7547217d667" />


| Model                         | loss_LeJEPA                         | loss_SIGReg                         | loss_Inv                           | acc_backbone_top1                  |
|------------------------------|-------------------------------------|-------------------------------------|------------------------------------|------------------------------------|
| resnet101 | $0.1217_{-0.0021}^{+0.0013}$         | $2.3796_{-0.0427}^{+0.0478}$         | $0.0755_{-0.0016}^{+0.0013}$        | $0.5626_{-0.0105}^{+0.0100}$        |
| resnet152 | $0.1175_{-0.0017}^{+0.0018}$         | $2.6001_{-0.0440}^{+0.0319}$         | $0.0669_{-0.0013}^{+0.0013}$        | $0.3507_{-0.0092}^{+0.0093}$        |
|resnet18  | $0.1497_{-0.0027}^{+0.0022}$         | $2.7681_{-0.0687}^{+0.0656}$         | $0.0960_{-0.0017}^{+0.0018}$        | $0.5820_{-0.0084}^{+0.0090}$        |
| resnet34  | $0.1373_{-0.0033}^{+0.0028}$         | $2.5768_{-0.0662}^{+0.0626}$         | $0.0872_{-0.0023}^{+0.0023}$        | $0.6209_{-0.0059}^{+0.0077}$        |
| resnet50  | $0.1277_{-0.0026}^{+0.0024}$         | $2.4180_{-0.0609}^{+0.0645}$         | $0.0809_{-0.0019}^{+0.0021}$        | $0.6594_{-0.0114}^{+0.0066}$        |
| vits      | $0.1491_{-0.0024}^{+0.0030}$         | $4.2122_{-0.0655}^{+0.1190}$         | $0.0659_{-0.0012}^{+0.0014}$        | $0.2508_{-0.0085}^{+0.0062}$        |
| vitt      | $0.1662_{-0.0024}^{+0.0024}$         | $4.3989_{-0.0728}^{+0.0796}$         | $0.0800_{-0.0015}^{+0.0014}$        | $0.1985_{-0.0050}^{+0.0059}$        |

**Comment by chemeris** [comment]
Hey, @ruiyicheng 

Have you tried replacing projectors with just the BatchNorm or manual batch normalization?

The original code uses `norm_layer=nn.BatchNorm1d` in the projector, which normalizes vectors before they reach SIGReg's ECF calculation. This may affect the SIGReg loss quite significantly.

It would be interesting to see if adding BatchNorm before SIGReg or a manual batch normalization into SIGReg calculation itself would help:
```python
        mu = proj.mean(dim=1, keepdim=True)
        sigma = proj.std(dim=1, keepdim=True).clamp_min(1e-6)
        proj = (proj - mu) / sigma
```

**Comment by ForestsKing** [comment]
@chemeris I've noticed this as well. The model's accuracy drops when BatchNorm is removed or replaced with LayerNorm.

**Comment by chemeris** [comment]
@ForestsKing Thank you, it's good to hear more confirmation!

ECF math assumes a centered distribution, and since we're calculating it over the B (batch) dimension, the distribution should be _batch_-normalized.

I've done a few experiments and ended up adding the normalization to the sigreg function unconditionally. It's a tiny overhead for the peace of mind.

That said, I don't think it's the only culprit with the projectors - they are bringing more than that, which I don't understand yet.

**Comment by RandallBalestriero** [comment]
Hello @chemeris  @ForestsKing ! Thanks for your messages! Let me clarify a few things and let me know if that helps or not
- *No centered assumption*: neither the ECF math nor SIGReg assumes a centered or scaled distribution. If you shift or scale the resulting ECF will encode that and produce a higher distance to the N(0,1) target distribution. I had experiments on that even in toy settings and we clearly see it reshift and rescale e.g. a ECF obtained from N(mu,sigma) with mu different than 0 and sigma different than 1
- *No normalization at last projector layer*: none of the projectors we used (neither in the paper nor in the minimal example) use a normalization layer at the last layer. We have BatchNorm in the hidden layers but the last one is a plain Linear layer

If you remove normalization and have issues learning as good of a representation it may just be an optimization issue (e.g. even in supervised learning, removing normalization layers from your model will result in trickier optimization and poorer performance in many cases). Lastly, the projector brings benefits by absorbing the bias from the incorrectly specified data augmentation, hyper-parameters etc, an example study looking at that is this one: https://arxiv.org/abs/2206.13378

Hope this helps!

**Comment by RandallBalestriero** [comment]
Closing this issue since you both gave the +1, feel free to reopen if needed!

**Comment by chemeris** [comment]
Interestingly, now with pure LeJEPA (V encoders with projectors, no predictor), removing normalization actually improves downstream task performance. This is quite the opposite of what I saw in the setup with a predictor.

<img width="675" height="527" alt="Image" src="https://github.com/user-attachments/assets/35b2ecb7-796d-46ff-a991-f78888fe143b" />

PS Thank you for the paper - very cool research, helps to improve my intuition (still reading).
